src/common/cockpitpipe.c

```diff
@@ -1525,6 +1525,37 @@ cockpit_pipe_skip (GByteArray *buffer,
   g_byte_array_remove_range (buffer, 0, skip);
 }
 
+gssize
+cockpit_pipe_parse_length (GByteArray *buffer,
+                           gsize *consumed)
+{
+  gchar *data;
+  guint32 i, size;
+
+  size = 0;
+  data = (gchar *)buffer->data;
+  for (i = 0; i < buffer->len; i++)
+    {
+      /* Check invalid characters, prevent integer overflow, limit max length */
+      if (i > 7 || data[i] < '0' || data[i] > '9')
+        break;
+      size *= 10;
+      size += data[i] - '0';
+    }
+
+  /* Want more data */
+  if (i == buffer->len)
+    return -1;
+
+  /* A failure */
+  if (data[i] != '\n')
+    return 0;
```

Comment:
The return value is a tristate. But I can change it around and document it better.